wasm/segment.go

```diff
@@ -8,104 +8,80 @@ import (
 	"github.com/tetratelabs/wazero/wasm/leb128"
 )
 
+// ImportKind indicates which import description is present
+// See https://www.w3.org/TR/wasm-core-1/#import-section%E2%91%A0
 type ImportKind = byte
 
 const (
-	ImportKindFunction ImportKind = 0x00
-	ImportKindTable    ImportKind = 0x01
-	ImportKindMemory   ImportKind = 0x02
-	ImportKindGlobal   ImportKind = 0x03
+	ImportKindFunc   ImportKind = 0x00
+	ImportKindTable  ImportKind = 0x01
+	ImportKindMemory ImportKind = 0x02
+	ImportKindGlobal ImportKind = 0x03
 )
 
-type ImportDesc struct {
-	Kind byte
-
-	// FuncTypeIndex is the index in Module.TypeSection corresponding to this import's FunctionType.
-	// This is only valid when Kind equals ImportKindFunction
-	FuncTypeIndex uint32
-	TableTypePtr  *TableType
-	MemTypePtr    *MemoryType
-	GlobalTypePtr *GlobalType
+// Import is the binary representation of an import indicated by Kind
+// See https://www.w3.org/TR/wasm-core-1/#binary-import
+type Import struct {
+	Kind ImportKind
+	// Module is the possibly empty primary namespace of this import
+	Module string
+	// Module is the possibly empty secondary namespace of this import
+	Name string
+	// DescFunc is the index in Module.TypeSection when Kind equals ImportKindFunc
+	DescFunc uint32
+	// DescTable is the inlined TableType when Kind equals ImportKindTable
+	DescTable *TableType
+	// DescMem is the inlined MemoryType when Kind equals ImportKindMemory
+	DescMem *MemoryType
+	// DescGlobal is the inlined GlobalType when Kind equals ImportKindGlobal
+	DescGlobal *GlobalType
 }
 
-func readImportDesc(r io.Reader) (*ImportDesc, error) {
+func readImport(r io.Reader) (i *Import, err error) {
+	i = &Import{}
+	if i.Module, err = readNameValue(r); err != nil {
+		return nil, fmt.Errorf("error decoding import module: %w", err)
+	}
+
+	if i.Name, err = readNameValue(r); err != nil {
+		return nil, fmt.Errorf("error decoding import name: %w", err)
+	}
+
 	b := make([]byte, 1)
-	if _, err := io.ReadFull(r, b); err != nil {
-		return nil, fmt.Errorf("read value kind: %v", err)
+	if _, err = io.ReadFull(r, b); err != nil {
+		return nil, fmt.Errorf("error decoding import kind: %w", err)
 	}
 
-	switch b[0] {
-	case ImportKindFunction:
-		tID, _, err := leb128.DecodeUint32(r)
-		if err != nil {
-			return nil, fmt.Errorf("read typeindex: %v", err)
+	i.Kind = b[0]
+	switch i.Kind {
+	case ImportKindFunc:
+		if i.DescFunc, _, err = leb128.DecodeUint32(r); err != nil {
+			return nil, fmt.Errorf("error decoding import func typeindex: %w", err)
 		}
-		return &ImportDesc{
-			Kind:          ImportKindFunction,
-			FuncTypeIndex: tID,
-		}, nil
 	case ImportKindTable:
-		tt, err := readTableType(r)
-		if err != nil {
-			return nil, fmt.Errorf("read table type: %v", err)
+		if i.DescTable, err = readTableType(r); err != nil {
+			return nil, fmt.Errorf("error decoding import table desc: %w", err)
 		}
-		return &ImportDesc{
-			Kind:         ImportKindTable,
-			TableTypePtr: tt,
```

Comment:
ps one of the motivations for flattening was reducing the amount of things we have to name. ex now we don't need to say things like Desc.TableTypePtr